natchez-fs2/src/main/scala/com/ovoenergy/effect/natchez/AllocatedSpan.scala

```diff
@@ -0,0 +1,112 @@
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+import fs2.concurrent.Queue
+import fs2.{Pipe, Stream}
+import natchez.{Kernel, Span, TraceValue}
+
+/**
+ * A Natchez span that has been pre-allocated and will stay open
+ * until either the stream that created it terminates or you call submit
+ */
+trait AllocatedSpan[F[_]] extends Span[F] {
+
+  /**
+   * Add a task to run on calling submit. The added task will run before the span is submitted.
+   * If it fails any errors will be discarded and the span will then be submitted
+   */
+  def addSubmitTask(task: F[Unit]): AllocatedSpan[F]
+
+  /**
+   * Submit the span. The span should be considered invalid after this is called
+   * and no subspans should be created
+   */
+  def submit: F[Unit]
+}
+
+object AllocatedSpan {
+
+  /**
+   * Given a span broken out of a natchez resource
+   * create an AllocatedSpan that allows us to submit it
+   */
+  private def createSpan[F[_]](
+    spn: Span[F], submitTask: F[Unit]
+  )(implicit F: Sync[F]): AllocatedSpan[F] =
+    new AllocatedSpan[F] {
+      def kernel: F[Kernel] =
+        spn.kernel
+      def put(fields: (String, TraceValue)*): F[Unit] =
+        spn.put(fields:_*)
+      def span(name: String): Resource[F, Span[F]] =
+        spn.span(name)
+      def addSubmitTask(task: F[Unit]): AllocatedSpan[F] =
+        createSpan(spn, F.uncancelable(F.attempt(task) >> submit))
+      def submit: F[Unit] =
+        submitTask
+    }
+
+  /**
+   * Associate a value with an allocated span
+   */
+  case class Traced[F[_], A](
+    span: AllocatedSpan[F],
+    value: A
+  )
+
+  object Traced {
+    implicit def traverse[F[_]]: Traverse[Traced[F, *]] =
+      cats.derived.semi.traverse[Traced[F, *]]
+  }
+
+  /**
+   * Create an AllocatedSpan which breaks a span out of a Natchez resource
+   * and instead allows it to be submitted manually when committing Kafka messages
+   */
+  def create[F[_]: Concurrent, A](maxOpen: Int = 100)(
+    rootSpan: A => Resource[F, Span[F]]
+  )(implicit F: Sync[F]): Pipe[F, A, Traced[F, A]] = { stream =>
+    /*
+     * First we create an outer queue that receives tasks to run and
+     * passes them along to parEvalMap so that interruption of the outer stream we're piping
+     * will cause all the tasks to be cancelled
+     */
+    Stream
+      .eval(Queue.bounded[F, F[Unit]](maxSize = 1))
+      .flatMap { taskQueue =>
+        stream
+          .concurrently(taskQueue.dequeue.parEvalMapUnordered(maxOpen)(identity))
+          .evalMap { item =>
+            /*
+             * We then create two queues - one to dig the Span out of the resource and another
+             * that we use to block the resource from quitting until we say so,
+             * either with an error or a unit to indicate success
+             */
+            for {
+              out <- Queue.bounded[F, Span[F]](maxSize = 1)
+              halt <- Queue.bounded[F, Either[Throwable, Unit]](maxSize = 1)
```

Comment:
I think I get now how this mechanism works - good idea.

I think you could potentially use lower-level primitives from cats-effect to achieve the same effect. `Deferred` should be sufficient here since you only complete those queues with one element and `Deferred.get` waits when it's empty.

Reply:
That's a good point. I wonder if an implementation with `Deferred` would look more intuitive.

Reply:
I tried to come up with a solution using `Deferred`, but I couldn't get it work. Did any of you have better luck?